### PR TITLE
PT-358/PT-359: New classic template

### PR DESCRIPTION
**Pull request by florianmski**
> Tracked in JIRA by [PT-359](https://novoda.atlassian.net/browse/PT-359)

### Scope of the PR

This PR introduced a redesigned classic widget. We now use flexbox and have a more minimal approach which allow us to display more  and bigger informations as well as a logo for context.

### Screenshots

| Before | After |
| --- | --- |
| webdashboard_before | webdashboard_after |

### CODE REVIEW

File: angular2-client/src/app/dashboards/classic/classic.component.scss [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+.content > span {
+  font-size: 10vmin;
+  font-family: 'Roboto';
+  max-height: 100%;
+  flex: 1 0;
+  overflow: hidden;
+  display: -webkit-box;
+  -webkit-line-clamp: 3;
+  -webkit-box-orient: vertical;
```
Comment (florianmski) [inline angular2-client/src/app/dashboards/classic/classic.component.scss]: Hacks to have an ellipsis for the text.
I chose max 3 lines but that could have been 5, I think it's also good to make sure we have something glancable and not too long

File: angular2-client/src/app/dashboards/classic/classic.component.ts [hunk 2]
```diff
@@ -1,18 +1,33 @@
...
+    this.configService = configService;
+  }
+
   public update(event: WidgetEvent) {
-    this.data = event.payload;
+    if (event.payload.serverImage) {
+      this.configService.getServerUrl()
+        .subscribe((serverUrl: String) => {
+          let payload = event.payload;
+          payload.serverImage = serverUrl + payload.serverImage;
```
Comment (florianmski) [inline angular2-client/src/app/dashboards/classic/classic.component.ts]: We need to get the server url if we want to get the image from there

File: angular2-client/src/app/dashboards/github-contributors/contributors-dashboard/contributors-dashboard.component.scss [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
     width: 100%;
   }
 
   .team-contributors, .project-contributors {
-    color: white;
+    color: black;
+    font-size: 4em;
     text-align: center;
     margin-top: 10px;
     margin-bottom: 0;
```
Comment (florianmski) [inline angular2-client/src/app/dashboards/github-contributors/contributors-dashboard/contributors-dashboard.component.scss]: Make everything bigger for the github contributors widget

File: node-server/index.js [hunk 4]
```diff
@@ -13,10 +16,15 @@ const dashboards = new Dashboards();
...
+  {
+    dataStore: new MemoryDataStore()
+  }
 );
 CONFIG.services.slack.slacker = new Slacker(rtmClient);
 const widgetLoader = new WidgetLoader(CONFIG.services, CONFIG.widgets, new ClassLoader());
 const server = new Server(widgetLoader, dashboards, clients);
 
-server.start(io, CONFIG.server.port);
+app.use('/public', express.static(__dirname + '/public'));
```
Comment (florianmski) [inline node-server/index.js]: We now serve static images to be used as context for the classic widget (logos of the services we get the data from for instance)

File: node-server/server/configs/sample.json [hunk 5]
```diff
@@ -28,8 +28,7 @@
   "reviews": {
     "apps": [
       {
-        "package": "com.app1",
-        "name": "App 1"
+        "package": "com.app1"
```
Comment (florianmski) [inline node-server/server/configs/sample.json]: We now get the name of the app directly for the API, less config 🎉

File: angular2-client/src/app/dashboards/classic/classic.component.html [hunk 6]
```diff
@@ -1,10 +1,5 @@
...
-  <div class="active-container">
-    <div class="active-filler"></div>
-    <span class="active-channel-header" [innerHTML]="data.title"></span>
-    <br>
-    <div>
-      <span class="active-channel-copy" [innerHTML]="data.text"></span>
-    </div>
-  </div>
+<div class="container">
+  <div class="header"><b><span [innerHTML]="data.title"></span></b></div>
```
Comment (eduardb) [inline angular2-client/src/app/dashboards/classic/classic.component.html]: Why using `<b>` here, and not making it bold from CSS?

File: angular2-client/src/app/dashboards/classic/classic.component.scss [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+.header {
+  flex: 1;
+  display: flex;
+  align-items: center;
+  justify-content: flex-start;
+  padding: 10pt 25pt;
+}
+
+.header span {
+  font-size: 8vmin;
```
Comment (eduardb) [inline angular2-client/src/app/dashboards/classic/classic.component.scss]: 👍 Nice, I also discovered `vmin` and `vmax` a few days ago.

File: angular2-client/src/app/dashboards/classic/classic.component.scss (hunk 1 above)
Comment (eduardb) [inline angular2-client/src/app/dashboards/classic/classic.component.scss]: It would have been interesting to make the text fill a specific part of the container, but I guess that the glanceability is more important 👍

File: node-server/index.js [hunk 8]
```diff
@@ -1,4 +1,7 @@
-const io = require('socket.io');
+const app = require('express')();
+const express = require('express');
```
Comment (eduardb) [inline node-server/index.js]: Express is back 😢 We could have used it for the "refresh" functionality. Oh, well ¯\\_(ツ)_/¯

File: node-server/server/file-resolver.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
+class FileResolver {
+
+  constructor(basePath) {
+    if (!basePath.startsWith('/') || basePath.endsWith('/')) {
+      throw new Error('basePath for FileResolver must start with a "/" and ends without');
```
Comment (eduardb) [inline node-server/server/file-resolver.js]: `ends` should be `end` 😛

File: node-server/server/widget/calendar.js [hunk 10]
```diff
@@ -12,21 +12,20 @@ class Calendar extends BaseWidget {
     this.client = new google.auth.JWT(
       googleConfig.client_email,
       null,
-      googleConfig.private_key,
-      ['https://www.googleapis.com/auth/calendar.readonly'],
+      googleConfig.private_key, ['https://www.googleapis.com/auth/calendar.readonly'],
```
Comment (eduardb) [inline node-server/server/widget/calendar.js]: Why this change? 😕

File: node-server/server/widget/calendar.js [hunk 11]
```diff
@@ -12,21 +12,20 @@ class Calendar extends BaseWidget {
...
       googleConfig.user
     );
   }
 
   rule() {
     return this._authorize(this.client)
-            .then(this._fetchNextEvent(this.client, this.config))
-            .then(this._toRuleResult.bind(this));
+      .then(this._fetchNextEvent(this.client, this.config))
```
Comment (eduardb) [inline node-server/server/widget/calendar.js]: 👍

File: node-server/server/widget/stackoverflow.js [hunk 12]
```diff
@@ -14,7 +14,7 @@ class StackOverflow extends BaseWidget {
   rule() {
     if (this.query) {
       return this._getQuestions()
-        .then(this._toRuleResult);
+        .then(data => this._toRuleResult(data));
```
Comment (eduardb) [inline node-server/server/widget/stackoverflow.js]: Again, why the change here? 👅

File: node-server/server/widget/stackoverflow.spec.js [hunk 13]
```diff
@@ -10,7 +10,9 @@ describe('Stackoverflow', () => {
   });
 
   it('should return correct payload', (done) => {
-    stackoverflow = new Stackoverflow({}, {query: 'novoda'}, mockHttpClient.request());
+    stackoverflow = new Stackoverflow({}, {
+      query: 'novoda'
+    }, mockHttpClient.request());
```
Comment (eduardb) [inline node-server/server/widget/stackoverflow.spec.js]: Shouldn't you also put the `request()` call on the next line?

File: node-server/server/widget/stackoverflow.spec.js [hunk 14]
```diff
@@ -61,7 +66,9 @@ describe('Stackoverflow', () => {
 
   it('should append query to request', (done) => {
     const httpClient = jasmine.createSpy().and.callFake(mockHttpClient.request());
-    stackoverflow = new Stackoverflow({}, {query: 'novoda'}, httpClient);
+    stackoverflow = new Stackoverflow({}, {
+      query: 'novoda'
+    }, httpClient);
```
Comment (eduardb) [inline node-server/server/widget/stackoverflow.spec.js]: Here as well

File: node-server/server/widget/stackoverflow.spec.js [hunk 15]
```diff
@@ -82,7 +89,12 @@ describe('Stackoverflow', () => {
   it('should append min date to request when in_the_last is defined', (done) => {
     const httpClient = jasmine.createSpy().and.callFake(mockHttpClient.request());
     const moment = jasmine.createSpy().and.returnValue(require('moment')(691804800000));
-    stackoverflow = new Stackoverflow({}, {query: 'novoda', in_the_last: {days: 1}}, httpClient, moment);
+    stackoverflow = new Stackoverflow({}, {
+      query: 'novoda',
+      in_the_last: {
+        days: 1
+      }
+    }, httpClient, moment);
```
Comment (eduardb) [inline node-server/server/widget/stackoverflow.spec.js]: and here

File: angular2-client/src/app/dashboards/classic/classic.component.scss (hunk 1 above)
Comment (florianmski) [inline angular2-client/src/app/dashboards/classic/classic.component.scss]: Believe me, I tried. Every "solution" were super hacky and none actually worked

Comment (frapontillo) [inline angular2-client/src/app/dashboards/classic/classic.component.scss]: I believe you. 😆

File: node-server/index.js (hunk 8 above)
Comment (florianmski) [inline node-server/index.js]: Yeah, still can change it ;)

File: node-server/server/widget/calendar.js (hunk 10 above)
Comment (florianmski) [inline node-server/server/widget/calendar.js]: That's an extension that will format on save, will revert.
I deactivated recently cause it was more annoying that it was useful :/

File: node-server/server/widget/stackoverflow.js (hunk 12 above)
Comment (florianmski) [inline node-server/server/widget/stackoverflow.js]: It's the good ol' `this` issue, we now need it in the `_toRuleResult`

File: angular2-client/src/app/dashboards/classic/classic.component.scss [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+.footer {
+  flex: 1;
+  display: flex;
+  align-items: center;
+  justify-content: flex-end;
+  padding: 10pt 25pt;
+}
+
+.footer > img {
```
Comment (florianmski) [inline angular2-client/src/app/dashboards/classic/classic.component.scss]: Neat! Didn't know
